File: scripts/check_references.js
#!/usr/bin/env node

/**
 * References Checker
 * Scans all markdown files in the repository to ensure all local markdown links resolve correctly.
 */

const fs = require('fs');
const path = require('path');

const rootDir = path.resolve(__dirname, '..');
let checkedCount = 0;
let brokenCount = 0;

function checkMarkdownFile(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const fileDir = path.dirname(filePath);

  // Match markdown links [text](path)
  const linkRegex = /\[.*?\]\((?!http|https|#|mailto:)(.*?)\)/g;
  let match;

  while ((match = linkRegex.exec(content)) !== null) {
    let target = match[1].split('#')[0]; // strip anchor
    if (!target) continue;

    // Handle file:/// absolute paths
    if (target.startsWith('file:///')) {
      target = target.replace(/^file:\/\/\//, '');
      // Handle windows paths e.g. c:/...
      if (process.platform === 'win32' && /^[a-zA-Z]:\//.test(target)) {
        target = target.replace(/\//g, '\\');
      }
    } else {
      target = path.resolve(fileDir, target);
    }

    checkedCount++;
    if (!fs.existsSync(target)) {
      console.error(`❌ Broken link in ${path.relative(rootDir, filePath)}: -> ${match[1]}`);
      brokenCount++;
    }
  }
}

function scanDir(dir) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.name === 'node_modules' || entry.name === '.git') continue;
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      scanDir(fullPath);
    } else if (entry.isFile() && entry.name.endsWith('.md')) {
      checkMarkdownFile(fullPath);
    }
  }
}

console.log('\n🔍 Checking all markdown references in repository...');
scanDir(rootDir);

if (brokenCount === 0) {
  console.log(`✅ REFERENCE CHECK PASS: All ${checkedCount} local links verified successfully!\n`);
  process.exit(0);
} else {
  console.log(`\n❌ REFERENCE CHECK FAILED: Found ${brokenCount} broken link(s) across ${checkedCount} checked.\n`);
  process.exit(1);
}

File: scripts/check_references.py
#!/usr/bin/env python3
"""
References Checker: Scans all markdown files to verify local links.
"""

import os
import sys
import re

root_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
checked_count = 0
broken_count = 0

def check_markdown_file(file_path):
    global checked_count, broken_count
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()

    file_dir = os.path.dirname(file_path)
    matches = re.findall(r"\[.*?\]\((?!http|https|#|mailto:)(.*?)\)", content)

    for target in matches:
        clean_target = target.split("#")[0]
        if not clean_target:
            continue

        if clean_target.startswith("file:///"):
            resolved = clean_target[8:]
            if sys.platform == "win32" and re.match(r"^[a-zA-Z]:\/", resolved):
                resolved = resolved.replace("/", "\\")
        else:
            resolved = os.path.normpath(os.path.join(file_dir, clean_target))

        checked_count += 1
        if not os.path.exists(resolved):
            print(f"❌ Broken link in {os.path.relpath(file_path, root_dir)}: -> {target}")
            broken_count += 1

def scan_dir(dir_path):
    for root, dirs, files in os.walk(dir_path):
        if "node_modules" in root or ".git" in root:
            continue
        for file in files:
            if file.endswith(".md"):
                check_markdown_file(os.path.join(root, file))

if __name__ == "__main__":
    print("\n🔍 Checking all markdown references in repository...")
    scan_dir(root_dir)

    if broken_count == 0:
        print(f"✅ REFERENCE CHECK PASS: All {checked_count} local links verified successfully!\n")
        sys.exit(0)
    else:
        print(f"\n❌ REFERENCE CHECK FAILED: Found {broken_count} broken link(s) across {checked_count} checked.\n")
        sys.exit(1)
